proposals/mpi-operator-proposal.md

```diff
@@ -0,0 +1,224 @@
+## Motivation
+Kubeflow currently supports distributed training of TensorFlow models using
+[tf-operator](https://github.com/kubeflow/tf-operator), which relies on
+centralized parameter servers for coordination between workers. An alternative
+is a decentralized approach in which workers communicate with each other
+directly via the MPI `allreduce` primitive, without using parameter servers.
+NVIDIA's [NCCL](https://developer.nvidia.com/nccl) library has implemented most
+of the MPI primitives efficiently on GPUs, and Uber's
+[Horovod](https://github.com/uber/horovod) makes it simple to do multi-gpu and
+multi-node training using TensorFlow; the TensorFlow team is also working on an
+`allreduce`
+[implementation](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/distribute).
+Compared to parameter servers, this approach arguably is more
+[bandwidth optimal](http://www.cs.fsu.edu/~xyuan/paper/09jpdc.pdf) and
+[scales better](https://eng.uber.com/horovod/).
+
+However, right now there is no easy way to launch this type of training jobs on
+Kubernetes. By providing a CRD and a custom controller, we can make
+allreduce-style distributed training as simple as training on a single node.
+
+## Goals
+* Provide a common Custom Resource Definition (CRD) for defining a single-gpu,
+multi-gpu, or multi-node training job.
+* Implement a custom controller to manage the CRD, create dependent resources,
+and reconcile the desired states.
+* The cross-pod communication should be secure, without granting unnecessary
+permissions to any pod.
+* Though the initial version focuses on TensorFlow/Horovod, the approach can be
+applied to other frameworks with MPI support, such as CNTK.
+
+## Non-Goals
+In theory, this operator can be used to run arbitrary MPI jobs (e.g.
+computational fluid dynamics), but it's not our focus.
+
+## API
+
+### Container Image
+
+A requirement on user's container image is that it includes an MPI
+implementation. The initial version is tested on Open MPI 3.0.0, but it can be
+extended to other MPI implementations if there is a need.
+
+It's also expected that the user would invoke `mpirun`, either through an
+`entrypoint` in the Docker image, or a `command` in the `PodSpec`.
+
+SSH is not needed (or used).
+
+### Custom Resource Definition
+The custom resource can be defined in two ways, in terms of how GPU resources
+are specified.
+
+In the simple version, user specifies the total number of GPUs and the operator
+figures out how to allocate them efficiently:
+```yaml
+apiVersion: kubeflow.org/v1alpha1
+kind: MPIJob
+metadata:
+  name: tensorflow-benchmarks
+spec:
+  gpus: 128
+  template:
+    spec:
+      containers:
+      - name: tensorflow-benchmarks
+        image: rongou/tensorflow_benchmarks:latest
+```
```

Comment:
Can you add example `mpirun` command here?  Can it be templated to automatically substitute # of allocated GPUs?

Reply:
OK, I added the `mpirun` command override in the "flexible" version down below. In the hostfile we create the number of slots are set to the number of GPUs requested, so you don't need to worry about # GPUs if you just do `-map-by slot`.